_posts/2023-06-30-types-and-programming-languages.markdown

```diff
@@ -0,0 +1,534 @@
+**Not being stuck** means that we will always be able to find a term to "replace" or to "compute", it's a bit of a synonym here.
+
+There are terms that we know we won't be able to do any replacement on them: Value terms.
+Here is what we want them to look like: Either Positive numbers, zero, negative numbers, or booleans.
+
+{% highlight javascript %}
+predicate IsSuccs(term: Term) {
+  term == Zero || (term.Succ? && IsSuccs(term.e))
+}
+predicate IsPreds(term: Term) {
+  term == Zero || (term.Pred? && IsPreds(term.e))
+}
+predicate IsFinalValue(term: Term) {
+  term == True || term == False || term == Zero ||
+  IsSuccs(term) || IsPreds(term)
+}
+{% endhighlight %}
+
+Now, we can write our one-step evaluation method. As a requirement, we add that the term must be well-typed and is not a final value.
+
+{% highlight javascript %}
+function OneStepEvaluate(e: Term): (r: Term)
+  requires WellTyped(e) && !IsFinalValue(e)
+{
+  match e {
+    case Succ(Pred(x)) =>
+      x
+
+    case Succ(succ) => 
+      Succ(OneStepEvaluate(succ))
+
+    case Pred(Succ(x)) =>
+      x
+
+    case Pred(pred) =>
+      Pred(OneStepEvaluate(pred))
+
+    case IsZero(Zero) =>
+      True
+
+    case IsZero(e) =>
+      if IsFinalValue(e) then
+        assert e.Pred? || e.Succ?;
+        False
+      else
+        IsZero(OneStepEvaluate(e))
+  
+    case Add(a, b) =>
+      if !IsFinalValue(a) then
+        Add(OneStepEvaluate(a), b)
+      else if !IsFinalValue(b) then
+        Add(a, OneStepEvaluate(b))
+      else if a.Zero? then
+        b
+      else if b.Zero? then
+        a
+      else if a.Succ? then
+        Succ(Add(a.e, b))
+      else
+        assert a.Pred?;
+        Pred(Add(a.e, b))
+
+    case Double(a) =>
+      if IsFinalValue(a) then
+        Add(a, a)
+      else
+        Double(OneStepEvaluate(a))
+
+    case If(cond, thn, els) =>
+      if IsFinalValue(cond) then
+        assert GetType(cond) == Some(Bool);
+        assert cond == True || cond == False;
+        if cond == True then
+          thn
+        else
+          assert cond == False;
+          els
+      else
+        If(OneStepEvaluate(cond), thn, els)
+  }
+}
+{% endhighlight %}
+
+The interesting points to note on the function above, in a language like Dafny where every pattern must be exhaustive, are the following:
+
+* Every call consist either of `OneStepEvaluate` on one sub-argument, or a transformation that reduces the size of the tree. So something is always happening here.
+* All the cases are covered, Dafny does not complain!
+  * For example, when encountering the case `IsZero(e)`, if `e` is a final value, we know it's either `Pred` or `Succ`. It cannot be `True` or `False` due to type checking, so we can conclude.
+  * Similarly, if the condition of an if term is a final value, because of type-checking, Dafny knows it's either True or False.
```

Comment:
```suggestion
  * Similarly, if the condition of an if term is a final value, because it's well-typed, Dafny knows it's either `True` or `False`.
```